docs/tutorial.md

```diff
@@ -3,200 +3,490 @@
+
+### Callbacks
+A [callback](https://en.wikipedia.org/wiki/Callback_(computer_programming)) is a function that you can give to another function, that is expected to be *called back*, generally when that function ends.  
+In Mineflayer, callbacks are often used to handle errors.
+
+```js
+bot.consume((error) => {
+  if (error) {
+    console.log(error)
+  } else {
+    console.log('Finished consuming')
   }
 })
 ```
 
+The above code will try to consume what the bot is currently holding.  
+When the consuming ends, the function that is passed along is called.  
+We can then do other things that we want to do after.  
+The function could also be called when an error occurs.
+
 ## Advanced
 
-For the following, we assume you have understood the [Basics](#basics) tutorial.
+The following concepts aren't necessary to create a Mineflayer bot, but they can be useful to understand and create more advanced bots.  
+We assume you have understood the [Basics](#basics) tutorial.
+
+### Asynchronousy
+In Javascript, asynchronousy is an important concept.  
+By default, Javascript will run everything line by line, and only go to the next line if the current line is done. This is called blocking.  
+However, sometimes you have to do something that takes a relatively long time, and you don't want your whole program to block and wait for it to finish.  
+
+Interacting with the filesystem is often done using asynchronousy, because reading and writing large files can take a long time.  
+
+```js
+const myPromise = new Promise((resolve, reject) => {
+  setTimeout( () => {
+    resolve("Success!")  // Yay! Everything went well!
+  }, 1000) 
+}) 
+
+myPromise.then((successMessage) => {
+  console.log(successMessage) 
+})
+
+myPromise.catch((failureMessage) => {
+  console.log(failureMessage)
+})
+```
+
+The above codes uses what is called a Promise. A promise promises it will eventually complete.  
+The function given you a promise always has 2 parameters, a `resolve` function and a `reject` function.  
+If the promise is successful, it will call the `resolve` function, otherwise it will call the `reject` function.  
+The above code uses a `setTimeout`, which calls the given function after the set amount of milliseconds, 1000 in this case.  
+You can then tell the promise what it should do when it succeeds with `.then(function)` or when it fails with `.catch(function)`
+
+The `.then` and `.catch` function can also be chained together with the promise to simplify the code.
+
+```js
+const myPromise = new Promise((resolve, reject) => {
+  setTimeout( () => {
+    resolve("Success!")  // Yay! Everything went well!
+  }, 1000) 
+}).then((successMessage) => {
+  console.log(successMessage) 
+}).catch((failureMessage) => {
```

Comment:
it should be `error` instead of `failureMessage`

Reply:
Done